### PR TITLE
Proper construction/destruction of JobInfo class

### PR DESCRIPTION
**Pull request by HereThereBeDragons**
Issue #3095 
Supersedes #3221 

and includes unittests for sinks

### CODE REVIEW

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by HereThereBeDragons** [comment]
@cernvm-bot help

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cpplint

**Comment by HereThereBeDragons** [comment]
@cernvm-bot tidy

**Comment by HereThereBeDragons** [comment]
@cernvm-bot unittest

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest cc8

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by vvolkl** [comment]
@cernvm-bot cloudtest

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

File: cvmfs/network/jobinfo.h [hunk 1]
```diff
@@ -33,106 +33,197 @@ namespace download {
...
-
-  /**
-   * Tells whether the error is because of a non-existing file. Should only
-   * be called if error_code is not kFailOk
-   */
-  bool IsFileNotFound();
+  off_t range_offset_;
+  off_t range_size_;
 
   // Internal state, don't touch
```
Comment (jblomer) [inline cvmfs/network/jobinfo.h]: It's already private
```suggestion
  // Internal state
```

File: cvmfs/network/jobinfo.h [hunk 2]
```diff
@@ -33,106 +33,197 @@ namespace download {
...
+  unsigned char num_retries_;
+  unsigned backoff_ms_;
+  unsigned int current_host_chain_index_;
+
+ public:
+  static JobInfo* CreateWithSink(const std::string *u, const bool c,
+                                 const bool ph, const shash::Any *h,
+                                 cvmfs::Sink *s);
+
+  static JobInfo* CreateWithoutSink(const std::string *u, const bool ph);
```
Comment (jblomer) [inline cvmfs/network/jobinfo.h]: Let's keep two regular constructor overloads instead. The advantage, in this case, is that we avoid heap additional allocations.  We should also document the effect of the two constructors (GET vs HEAD request).

File: cvmfs/network/jobinfo.h [hunk 3]
```diff
@@ -33,106 +33,197 @@ namespace download {
...
-    if (pipe_job_results.IsValid()) {
-      pipe_job_results.Destroy();
-    }
+  /**
+   * Tells whether the error is because of a non-existing file. Should only
+   * be called if error_code is not kFailOk
+   */
+  bool IsFileNotFound();
+
+  // Getter Pointer
```
Comment (jblomer) [inline cvmfs/network/jobinfo.h]: remove

File: cvmfs/network/jobinfo.h [hunk 4]
```diff
@@ -33,106 +33,197 @@ namespace download {
...
+  z_stream *GetZstreamPtr() { return &zstream_; }
+  Failures *GetErrorCodePtr() { return &error_code_; }
+  void **GetCredDataPtr() { return &cred_data_; }
+  curl_slist **GetHeadersPtr() { return &headers_; }
+  CURL **GetCurlHandle() { return &curl_handle_; }
+  shash::ContextPtr *GetHashContextPtr() { return &hash_context_; }
+  Pipe<kPipeDownloadJobsResults> *GetPipeJobResultWeakRef() {
+                                           return pipe_job_results.weak_ref(); }
+
+  // Getter
```
Comment (jblomer) [inline cvmfs/network/jobinfo.h]: Remove

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by HereThereBeDragons** [comment]
for cloudtest failure also see https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1780/ which is build on top of this branch. 
there only [ServerIntegrationTests.Commit command in a repository connected to the gateway](https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsCloudTesting/1780/CVMFS_PLATFORM=debian11,CVMFS_PLATFORM_CONFIG=x86_64,label=trampoline/testReport/(root)/ServerIntegrationTests/Commit_command_in_a_repository_connected_to_the_gateway/) fails on debian

**Review by jblomer — CHANGES_REQUESTED** [review]
Very nice! We may need to come back to the `JobInfo` object at some point but this is strictly better than before!

One larger change is sticking to constructors rather than the factory methods, to avoid a bunch of heap allocations. This unfortunately triggers changes everywhere where `JobInfo` is used, but at least it should be a schematic change.

File: cvmfs/network/jobinfo.h (hunk 3 above)
Comment (HereThereBeDragons) [reply cvmfs/network/jobinfo.h]: i think we had this discussion before: its used in `repository.cc`

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by HereThereBeDragons** [comment]
@cernvm-bot help

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest cc7

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

**Comment by jblomer** [comment]
See clang-tidy reported defect: https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsTidy/label=docker-x86_64/915/artifact/tidy.out/*view*/

`void SetProxy(std::string proxy)` --> `void SetProxy(const std::string &proxy) `

**Review by jblomer — APPROVED** [review]
Very nice improvement!

**Comment by HereThereBeDragons** [comment]
@cernvm-bot cloudtest

File: test/unittests/t_sink.cc [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+  // recheck content
+  ASSERT_FALSE(sink.IsValid());
+  int val = unlink(dest_path.c_str());
+
+  ASSERT_EQ(val, -1);
+  ASSERT_EQ(errno, ENOENT);
+}
+
+TEST_F(T_Sink, PathSinkNotValid) {
+  cvmfs::FileSink sink(NULL);
```
Comment (vvolkl) [inline test/unittests/t_sink.cc]:
```suggestion
  cvmfs::PathSink sink(NULL);
```

**Review by vvolkl — CHANGES_REQUESTED** [review]
Looks good to me apart from the typo in the unittests.

**Comment by HereThereBeDragons** [comment]
@cernvm-bot  cloudtest

File: test/unittests/t_sink.cc (hunk 5 above)
Comment (HereThereBeDragons) [reply test/unittests/t_sink.cc]: good catch, thanks

**Pull request merged**